### PR TITLE
feat(pack): add support for scaffolding new extensions

### LINKED ISSUE #1636: pack should support `pack extension new`
### Description
This is a sub-task of https://github.com/buildpacks/pack/issues/1470

Just like scaffold new buildpacks, we should be able to scaffold new extensions

### Proposed solution
The command should work just like `pack buildpack new`, but with extensions:

```
Usage:
  pack buildpack new <id> [flags]

Examples:
pack buildpack new sample/my-buildpack

Flags:
  -a, --api string       Buildpack API compatibility of the generated buildpack (default "0.8")
  -h, --help             Help for 'new'
  -p, --path string      Path to generate the buildpack
  -s, --stacks strings   Stack(s) this buildpack will be compatible with
                         Repeat for each stack in order, or supply once by comma-separated list (default [io.buildpacks.stacks.bionic])
  -V, --version string   Version of the generated buildpack (default "1.0.0")
```

### Additional context
- [x] This feature should be documented somewhere

### PR DESCRIPTION
**Pull request by sarthaksarthak9**
add support for scaffolding new extensions

## Summary
add support for scaffolding new extensions

#### After
Screenshot from 2024-01-17 21-26-07

## Documentation

- Should this change be documented?
    - [x] Yes, see #___
    - [ ] No

## Related

Resolves #1636

### CODE REVIEW

**Comment by sarthaksarthak9** [comment]
@natalieparellano could you plz review the pr ?

File: internal/commands/extension_new.go [hunk 1]
```diff
@@ -15,21 +24,66 @@ type ExtensionNewFlags struct {
...
+				return fmt.Errorf("directory %s exists", style.Symbol(path))
+			}
+
+			var stacks []dist.Stack
+			for _, s := range flags.Stacks {
+				stacks = append(stacks, dist.Stack{
+					ID:     s,
+					Mixins: []string{},
+				})
+			}
```
Comment (natalieparellano) [inline internal/commands/extension_new.go:66]: We can remove this part, stacks are not relevant for extensions

File: internal/commands/extension_new.go [hunk 2]
```diff
@@ -15,21 +24,66 @@ type ExtensionNewFlags struct {
...
+
+			if err := creator.NewExtension(cmd.Context(), client.NewExtensionOptions{
+				API:     flags.API,
+				ID:      id,
+				Path:    path,
+				Stacks:  stacks,
```
Comment (natalieparellano) [inline internal/commands/extension_new.go:72]:
```suggestion
```

File: internal/commands/extension_new.go [hunk 3]
```diff
@@ -15,21 +24,66 @@ type ExtensionNewFlags struct {
...
 			return nil
 		}),
 	}
-
-	// flags will go here
+	cmd.Flags().StringVarP(&flags.API, "api", "a", "0.8", "Extension API compatibility of the generated buildpack")
+	cmd.Flags().StringVarP(&flags.Path, "path", "p", "", "Path to generate the extension")
+	cmd.Flags().StringVarP(&flags.Version, "version", "V", "1.0.0", "Version of the generated extension")
+	cmd.Flags().StringSliceVarP(&flags.Stacks, "stacks", "s", nil, "Stack(s) this buildpack will be compatible with"+stringSliceHelp("stack"))
+	cmd.Flags().MarkDeprecated("stacks", "prefer `--targets` instead: https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md")
```
Comment (natalieparellano) [inline internal/commands/extension_new.go:86]:
```suggestion
```

Comment (natalieparellano) [reply internal/commands/extension_new.go:86]: Can we add `--targets` here? Does that exist for `buildpack new`?

File: pkg/client/new_extension.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	Path string
+
+	// The ID of the output extension artifact.
+	ID string
+
+	// version of the output extension artifact.
+	Version string
+
+	// Deprecated: The stacks this extension will work with
+	Stacks []dist.Stack
```
Comment (natalieparellano) [inline pkg/client/new_extension.go:30]:
```suggestion
```

File: pkg/client/new_extension.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+func (c *Client) NewExtension(ctx context.Context, opts NewExtensionOptions) error {
+	err := createExtensionTOML(opts.Path, opts.ID, opts.Version, opts.API, c)
+	if err != nil {
+		return err
+	}
+	return createBashExtension(opts.Path, c)
+}
+
+func createBashExtension(path string, c *Client) error {
+	if err := createBinScript(path, "build", bashBinBuild, c); err != nil {
```
Comment (natalieparellano) [inline pkg/client/new_extension.go:42]:
```suggestion
	if err := createBinScript(path, "generate", bashBinGenerate, c); err != nil {
```

Comment (natalieparellano) [reply pkg/client/new_extension.go:42]: We could do something like the below for `bashBinGenerate`:

```
#!/usr/bin/env bash
set -eo pipefail

# 1. GET ARGS
output_dir=$CNB_OUTPUT_DIR

# 2. GENERATE build.Dockerfile
cat >> "${output_dir}/build.Dockerfile" <<EOL
ARG base_image
FROM \${base_image}

RUN echo "Hello from build extension"
EOL

# 3. GENERATE run.Dockerfile
cat >> "${output_dir}/run.Dockerfile" <<EOL
ARG base_image
FROM \${base_image}

RUN echo "Hello from run extension"
EOL
```

Comment (natalieparellano) [reply pkg/client/new_extension.go:42]: https://buildpacks.io/docs/extension-guide/create-extension/ contains more information for the curious...

File: internal/commands/extension_new.go [hunk 6]
```diff
@@ -15,21 +24,66 @@ type ExtensionNewFlags struct {
...
+				return err
+			}
+
+			logger.Infof("Successfully created %s", style.Symbol(id))
 			return nil
 		}),
 	}
-
-	// flags will go here
+	cmd.Flags().StringVarP(&flags.API, "api", "a", "0.8", "Extension API compatibility of the generated buildpack")
```
[the remark this answers is not in the document]
Comment (natalieparellano) [reply internal/commands/extension_new.go:82]: 0.9 is the earliest API that supports extensions: https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.9

File: internal/commands/extension_new.go (hunk 3 above)
Comment (sarthaksarthak9) [reply internal/commands/extension_new.go:86]: yes, they exist for buildpack new

Comment (sarthaksarthak9) [reply internal/commands/extension_new.go:86]: here, targets are not present in ExtensionNewFlags struct !! do I need to add them ?

**Comment by sarthaksarthak9** [comment]
@natalieparellano I'm working on test for newExtension. Do we have anything similar to newBuildpack() method https://pkg.go.dev/github.com/buildpacks/pack/internal/commands/testmocks#MockPackClientMockRecorder.NewBuildpack for extension ?

File: internal/commands/extension_new.go (hunk 3 above)
Comment (sarthaksarthak9) [reply internal/commands/extension_new.go]: @natalieparellano  also, targets are not present in ExtensionDescriptor struct !! https://github.com/buildpacks/pack/blob/main/pkg/dist/extension_descriptor.go

**Comment by sarthaksarthak9** [comment]
I run make build make format make tidy and make verify they are running good at my end !! I don't know y these tests are failing !!

**Comment by jjbustamante** [comment]
> I run make build make format make tidy and make verify they are running good at my end !! I don't know y these tests are failing !!

Hi @sarthaksarthak9 

For what I can see, it looks like your mock setup is missing some information

```bash
 Got: {0.8 /tmp/build-test1320509439/targets example/targets 1.0.0 [] [{linux arm v6 [{ubuntu [14.04 16.04]}]}]} (client.NewBuildpackOptions)

  Want: is equal to {0.8 /tmp/build-test1320509439/targets example/targets 1.0.0} (client.NewExtensionOptions)
```

Your expected object is missing ` [] [{linux arm v6 [{ubuntu [14.04 16.04]}]}]` these attributes

**Comment by sarthaksarthak9** [comment]
> > I run make build make format make tidy and make verify they are running good at my end !! I don't know y these tests are failing !!
> 
> Hi @sarthaksarthak9
> 
> For what I can see, it looks like your mock setup is missing some information
> 
> ```shell
>  Got: {0.8 /tmp/build-test1320509439/targets example/targets 1.0.0 [] [{linux arm v6 [{ubuntu [14.04 16.04]}]}]} (client.NewBuildpackOptions)
> 
>   Want: is equal to {0.8 /tmp/build-test1320509439/targets example/targets 1.0.0} (client.NewExtensionOptions)
> ```
> 
> Your expected object is missing ` [] [{linux arm v6 [{ubuntu [14.04 16.04]}]}]` these attributes.
> 
> Maybe here:
> 
> ```go
> mockClient.EXPECT().NewBuildpack(gomock.Any(), client.NewExtensionOptions{
> 					API:     "0.8",
> 					ID:      "example/targets",
> 					Path:    filepath.Join(tmpDir, "targets"),
> 					Version: "1.0.0",
> 				}).Return(nil).MaxTimes(1)
> ```
> 
> You need to add more details on the targets options

thank you

**Comment by sarthaksarthak9** [comment]
@jjbustamante targets are not present in ExtensionDescriptor struct !! https://github.com/buildpacks/pack/blob/main/pkg/dist/extension_descriptor.go

**Comment by sarthaksarthak9** [comment]
cc @natalieparellano @jjbustamante could you pls review the above changes!!

**Comment by sarthaksarthak9** [comment]
Screenshot from 2024-01-27 10-15-31
@natalieparellano @jjbustamante I think we have testDeepContains as last unit test, still Make file throwing an error!! Could you pls check this out!!

**Comment by jjbustamante** [comment]
Hi @sarthaksarthak9 

I think the problem is that your tests are still not passing. I cloned your branch and try to fix them. This is what I changed:

Screenshot 2024-01-29 at 10 55 52 AM

I think you copied and pasted some tests from a previous command, which is ok, but you didn't changed the actual method you wanna mock `NewExtension`. The other error seems to be with the API version.  

I did those changes and the tests started passing locally

**Comment by sarthaksarthak9** [comment]
> Hi @sarthaksarthak9
> 
> I think the problem is that your tests are still not passing. I cloned your branch and try to fix them. This is what I changed:
> 
> Screenshot 2024-01-29 at 10 55 52 AM
> 
> I think you copied and pasted some tests from a previous command, which is ok, but you didn't changed the actual method you wanna mock `NewExtension`. The other error seems to be with the API version.
> 
> I did those changes and the tests started passing locally

srry for this , I forget to update the file

**Comment by sarthaksarthak9** [comment]
> Hi @sarthaksarthak9
> 
> I think the problem is that your tests are still not passing. I cloned your branch and try to fix them. This is what I changed:
> 
> Screenshot 2024-01-29 at 10 55 52 AM
> 
> I think you copied and pasted some tests from a previous command, which is ok, but you didn't changed the actual method you wanna mock `NewExtension`. The other error seems to be with the API version.
> 
> I did those changes and the tests started passing locally

Thank you for this ... Yah, I miss to update it !!

**Comment by sarthaksarthak9** [comment]
I commits the changes with wrong author, due to which DCO is not passing because of this I am closing this pull request and opening a new

**Pull request closed without merging**